# DP: Fix PR c++/66857, taken from the trunk

gcc/cp/

2015-07-25  Patrick Palka  <ppalka@gcc.gnu.org>

	PR c++/66857
	* cvt.c (ocp_convert): Don't call scalar_constant_value when
	converting to a class type.

gcc/testsuite/

2015-07-25  Patrick Palka  <ppalka@gcc.gnu.org>

	PR c++/66857
	* g++.dg/init/pr66857.C: New test.

--- a/src/gcc/cp/cvt.c
+++ b/src/gcc/cp/cvt.c
@@ -687,7 +687,8 @@
     }
 
   /* FIXME remove when moving to c_fully_fold model.  */
-  e = scalar_constant_value (e);
+  if (!CLASS_TYPE_P (type))
+    e = scalar_constant_value (e);
   if (error_operand_p (e))
     return error_mark_node;
 
--- a/src/gcc/testsuite/g++.dg/init/pr66857.C
+++ b/src/gcc/testsuite/g++.dg/init/pr66857.C
@@ -0,0 +1,19 @@
+// PR c++/66857
+// { dg-do run }
+
+const int i = 0;
+
+struct Test
+{
+  Test (const int &rhs)
+  {
+    if (&rhs != &i)
+      __builtin_abort ();
+  }
+};
+
+int
+main (void)
+{
+  Test test = i;
+}
